### DISCUSSION TITLE
How to record and transmit 125kHz/175kHz and similar frequencies? (RX/TX)

### DISCUSSION THREAD

**Discussion by kenwawa01** [merbanan/rtl_433#2966] [Q&A]
Hi!
One thing which comes to my mind is to record the transmission between  tpms wakeup/programming tool and tpms sensor itsel. AFAIK it's based on low frequency transmission, usually probably 125kHz.
On the Internet I can find many many threads about 433MHz transmission analysis but it seems that noone (or almost noone) is having public discussion about this LF part.

1. What device would be the best for recording? I was wondering if such a device could help here: https://pl.aliexpress.com/item/1005003747296138.html If not, would you recommend?

2. Is there any chance to make rtl_433 (perhaps with some minor modifications) to work on 125kHz with this device? If not, what software could you recommend?
 
3. Have you ever found any projects publicly available when someone tried to analyse tpms wakeup/activation/programming sequences? All links would be highly welcome. :P

**Comment by gdt** [comment]
You'll need an antenna/sdr that can receive around the frequency of interest.   Then I expect rtl_433 is suitable.  But I really am not familiar with such SDRs.

I am unaware of TPMS wakeup/programming, but my impression is that wakeup is not sophisticated.

**Reply by kenwawa01** [reply]
Well, I didn't find to much information but it looks like that the structure of message is also quite complicated in some systems and similar to tpms messages.
But I would like to learn more and for that I need to catch radio signal. I tried with one rfid card reader but... I wasn't able to catch anything. Perhaps it would be possible if I have very low level radio data software reader. The software attached to those rfid readers expect some kind of structure for known rfid cards types.

**Reply by hornbyp** [reply]
> But I would like to learn more and for that I need to catch radio signal. I tried with one rfid card reader but... I wasn't able to catch anything. 

I've recently completed a project to capture the [PKES](https://www.google.co.uk/search?q=passive+keyless+entry+system) signals from a Hyundai Tucson (to prove that my key Fob security timeout is working). TPMS is next on my list :-)

The 125KHz signal is **not** a radio wave - it is a magnetic field.  (AIUI, to propagate as a Radio wave, it would need to emanate from a 1/4 wavelength aerial - which is pretty long for 125KHz!). To receive it, you need a simple air coil, in very close proximity. (Ready made version here: [https://www.rfsolutions.co.uk/rfid-coil-antenna-125khz-p66](https://www.rfsolutions.co.uk/rfid-coil-antenna-125khz-p66) for example). 

The signal can then be rectified (germanium or schottky  diode) and filtered via a simple parallel RC circuit. Feeding the resulting signal into a [logic analyser](https://hobbycomponents.com/testing/673-hobby-components-usb-8ch-24mhz-logic-analyser-bundle), will allow you to visualise the signal. Tuning the coil (with a parallel capacitor) will boost the input voltage signal significantly, but will alter the timing (because it will take time to build-up to its full value and will decay away slowly. See figure 5 in this document: [https://ww1.microchip.com/downloads/en/AppNotes/00232B.pdf](https://ww1.microchip.com/downloads/en/AppNotes/00232B.pdf). 
There is more related information in this Microchip document: [AN912](https://ww1.microchip.com/downloads/aemDocuments/documents/OTH/ApplicationNotes/ApplicationNotes/00912a.pdf)

**Comment by pswiatki** [comment]
@kenwawa01: You could use RTL-SDR in direct sampling mode. It allows reception from 0 to 14.4 MHz. Check: [this article](https://www.rtl-sdr.com/rtl-sdr-direct-sampling-mode/)

**Comment by kenwawa01** [comment]
I'm wondering if reasonable signal quality is here closer to 14 MHz or 0 MHz... Perhaps I will try with this aliexpress box. Probably it's basing on the same trick as in this article but I hope that slightly improved...?

However it will be only part of the game. As it will allow listening 125kHz. But how to transmit the same signal for short distances 10-20cm like in tpms activation tools?

**Comment by pswiatki** [comment]
No TX is mentioned in the title of this thread.

**Comment by kenwawa01** [comment]
Right! Improved! :)

**Comment by kenwawa01** [comment]
Let's focus first of reading 125kHz transmission.

Ok with Gqrx I was able to observe 125kHz transmission. It seems to be much slower/longer that 433 MHz as traces for 433MHz are very small height size and for 125kHz it seems to last few seconds (approx. 20x longer).

Currently I'm trying to use rtl_433 to decode it, but despite many attempts Im getting exactly 0. I tried to write my own small decoder and also to use -A function and in both cases I get exactly nothing.

Any suggestions what parameters should I set to rtl_433 to get this 125kHz transmission for continental tpms sensors.

My examples (sure I've did some experiments with -g values and -direct_samp=1/2 and not sure which -s would be the best, if I set it to lower values I get message about failing to set sample rate ):

```
rtl_433 -A -R 0 -T 20 -f 125000 -g 20 -s 250000 -t "direct_samp=1" -v -W test 

rtl_433 -S all -R 258 -T 20 -f 125000 -g 20 -s 7987 -t "direct_samp=1" -v -W test 
```

R=258 it's my own decoder.

Is some parameters combinations I also get the message about setting low frequency filter. But when i tried to modify it with -l parameter, rtl_433 doesn't run normally but just display message as below and exists:

`	Level limit has changed from "-l 1000" to "-Y level=-12.1" in dB.`

It's a bit surprising than when I set frequency to 125kHz, the low pas filter for 250kHz is set, ex. for this:

`rtl_433 -S all -R 258 -T 20 -f 125000 -g 20  -t "direct_samp=1" -v -W test`

I get:

`[Baseband] low pass filter for 250000 Hz at cutoff 2500 Hz, 40.0 us`

How to turn off the low pass filter?

Any suggestions would be welcome?

**Comment by kenwawa01** [comment]
Just to add, I tried to increase scan to 1024k and tried also with

-M level -Y autolevel -Y minmax

Btw, I'm using Gqrx to observe my signal. Is there any way to record this signal from Gqrx as some kind of audio file to observe how does it change exactly in time? Ex. to try to guess how long last 0 an 1. What ofther Linux based software could be helpful here?

**Comment by zuckschwerdt** [comment]
You don't need the filter options (`-Y filter=`), that's just for very fast signals. I guess the 125 kHz is direct sampling (without IF) so you might need a higher sample rate, `-s 1000k`.

Try to record the signal (`-f 125k -s 1000k -w thefile_0.125M_1000k.cu8 -T 10` for 10 seconds) and upload as zip here

**Comment by kenwawa01** [comment]
I feel a bit lost...

I've recorded cu8 files as you suggest, with direct sampling on and off and... for me these files look empty. It could also happen that it did sth wrong when importing them to Audacity as I received two times longer time than expected.

However I also recorded two files from Gqrx in order to have comparison - as raw files from this software and here the signals are clearly visible.

If I look on this in Qgrx I can see 3 signals, two very strong and 3rd hardly noticeable (probably the 3rd is the sensor response).

It might be also the case that the quality of my receiver is too low...

For rtl_433 I decided to use two samplings 500k and 1000k as I was afraid that low cut filter which is set automatically might some impact on 1000k results (it's set to 100kHz then).

Any comments from your side?

And the files:

rtl_433
https://drive.google.com/file/d/1E2xLuQV5kSmDMt_yHkx4FZZ8SymhzcyG/view?usp=sharing 
https://drive.google.com/file/d/1F8zZM7OEPzUXj6vUYVbh_jv3EqIYyqVd/view?usp=sharing 
https://drive.google.com/file/d/1FXXKPiQQrCPud_ANc_Luk9wo8dJC4aVW/view?usp=sharing 
https://drive.google.com/file/d/1IX_I2YxrZRWTMEIx2AedqNpb471Ud1Tx/view?usp=sharing
https://drive.google.com/file/d/1MG-S0BXv9ozzQGRIOR9f4ZDp6E2SzEPC/view?usp=sharing
https://drive.google.com/file/d/1lFj27JJiOUz9Tmxb8-dmbSnSREsDyYj3/view?usp=sharing
https://drive.google.com/file/d/1zUhRjNk3Q4Ppgp_CNwm0SuQ5TKCZwQiE/view?usp=sharing

Gqrx
https://drive.google.com/file/d/1MQPqIS4KWpVcjGshIy7dIb2VAHClikY1/view?usp=sharing
https://drive.google.com/file/d/1oQeeUopephJtB13Hu2xRQwQ-B9m95kH0/view?usp=sharing

And one thing more, when using the same device on 433MHz in rtl_433, I'm getting very clear signal without any problems.

**Reply by zuckschwerdt** [reply]
The files are just very quiet, not empty. And there is a signal at 125kHz, observe in https://triq.org/pdv/ (or https://triq.org/spectrogram-next/ for zoom-in and waveform display).
rtl_433 can't deal with signals that faint. I never tried direct sampling, but I guess offsetting the frequency could help, does `-f 100k` work any better and show the signal still at 125k?

**Comment by kenwawa01** [comment]
Well, I will have to spend more time on that (and time is missing).

I'm wondering about one thing. The signal from rtl_433 is so weak... but the signal from Gqrx is so strong and visible. Why is that like this? Have you looked on Gqrx files?

I will try with -f 100k later.

I'm also wondering if using any of arduino rfid readers 125kHz connected to arduino or esp32 could be a better idea here? Those "antennas" are for one frequency only and the signal should be of better quality...? Ex. sth like this https://pl.aliexpress.com/item/1005006337836266.html or this https://pl.aliexpress.com/item/1005002688104886.html

I searched a lot of web pages and I was not able to find any DIY homebrew project for waking up tpms sensors (not counting those the simplest/oldest type ones where almost any 125kHz without any data can wake them up)... Have you perhaps seen sth as example?

**Reply by zuckschwerdt** [reply]
The Gqrx data is demodulated. Is it a FM demod? Then in theory amplitude won't matter.

As @hornbyp commented the signal is magnetic, maybe that's not common or interesting in the SDR space.
I once tried to use a salvaged 125kHz coil, but reconnecting it demands re-tuning it, which I didn't even try.

**Comment by kenwawa01** [comment]
Well, in Gqrx data might be demodulated or not. I set it to Raw IQ mode. I will have to check how much does it differs from Demodulated off. But you are right, it might be the reason.

I might try one of these devices from aliexpress and compare how it works. They are not expensive...

Do you have any suggestion what might be the parameters of this 125kHz signal?

**Comment by kenwawa01** [comment]
Ok, I tried a lot of different option and for me my recordings are worth nothings. In my humble opinion the signal to noise ration is too bad and it's difficult to distinct noise from real signal. There is sth in those recordings but without any value.

I have to focus how to improve the quality of signal before I will do sth more...

What antenna would be good for 125kHz signal? Good and cheap sure. :P

**Reply by pswiatki** [reply]
¼λ ground plane. I use it - works perfectly on 433M.

**Reply by zuckschwerdt** [reply]
I expect 433MHz and 125 kHz to be quite different. For the later the magnetic field will be more useful, which suggests a tuned coil.

**Reply by pswiatki** [reply]
ohhh... dammit, I forgot!! This is not 433M. Disregard my previous post.

Having done radio circuits for fun with my father (some 45 years ago, or so, i.e. when I was a kid)  I would expect a ferrite antena from an old radio (with LW band) to work perfectly.
Bear in mind, though, it is directional and the best signal is from the direction perpendicular to the ferrite rod (and the coil).

**Reply by hornbyp** [reply]
> Ok, I tried a lot of different option and for me my recordings are worth nothings. In my humble opinion the signal to noise ration is too bad and it's difficult to distinct noise from real signal. 
> What antenna would be good for 125kHz signal? Good and cheap sure. :P

I refer you to my response of two weeks ago, which you didn't acknowledge.

A simple air coil (pre-fabricated or home made) should yield a signal of several **VOLTS** pk-pk. Stuffing that into your average RTL-SDR capable hardware would probably do it no good at all! (Can it even tune that low?)

Tuning the coil with a capacitor (or adding a ferrite rod) will increase the signal level - at the expense of the waveform shape (see that microchip PDF).  My experiments with tuning, used a 125KHz signal from a microcontroller and a 'scope. 

(The 125KHz transmitter is (probably) modulating the signal by turning it ON and OFF. However, its LC tank circuit takes a little while to 'fill up' and a little while to 'empty' when the signal is applied and removed. If you add more of this at the receiving end, it will be difficult to ascertain what the timing actually is).

If you only want to capture and reproduce this 125KHz signal, you don't actually _need_ to **understand** it! The PKES signal that emanates from my car door, [looks like this: ](https://triq.org/pdv/#AAB02F06010084017C00FC0884020C000080808080808080808080809180A2A28280A280A080808280A280A2A280A2A0828455+AAB00F06010084017C00FC0884020C0000B555)

I can see a Preamble, a Start-bit, Data and a Stop bit. Effectively, the data is represented by four combinations of short or long pulses and short or long gaps, so that's what my software recreates - with no idea of its meaning. (If anyone can spot the encoding, I _would_ be interested to know!)

**Reply by hornbyp** [reply]
This is an example of the sort of received signal you should see - in this case using about 30 turns of flexible wire wound into a ~30cm diameter coil.

![ReceivedSignals](https://github.com/merbanan/rtl_433/assets/17692861/a2f98b3d-ff72-443a-a95a-3d65293f6b0a)

The blue signal is the data you're trying to recover, (modulating a 125KHz square wave using a mosfet H bridge)

**Reply by fibunaci** [reply]
> > Ok, I tried a lot of different option and for me my recordings are worth nothings. In my humble opinion the signal to noise ration is too bad and it's difficult to distinct noise from real signal.
> > What antenna would be good for 125kHz signal? Good and cheap sure. :P
> 
> I refer you to my response of two weeks ago, which you didn't acknowledge.
> 
> A simple air coil (pre-fabricated or home made) should yield a signal of several **VOLTS** pk-pk. Stuffing that into your average RTL-SDR capable hardware would probably do it no good at all! (Can it even tune that low?)
> 
> Tuning the coil with a capacitor (or adding a ferrite rod) will increase the signal level - at the expense of the waveform shape (see that microchip PDF). My experiments with tuning, used a 125KHz signal from a microcontroller and a 'scope.
> 
> (The 125KHz transmitter is (probably) modulating the signal by turning it ON and OFF. However, its LC tank circuit takes a little while to 'fill up' and a little while to 'empty' when the signal is applied and removed. If you add more of this at the receiving end, it will be difficult to ascertain what the timing actually is).
> 
> If you only want to capture and reproduce this 125KHz signal, you don't actually _need_ to **understand** it! The PKES signal that emanates from my car door, [looks like this: ](https://triq.org/pdv/#AAB02F06010084017C00FC0884020C000080808080808080808080809180A2A28280A280A080808280A280A2A280A2A0828455+AAB00F06010084017C00FC0884020C0000B555)
> 
> I can see a Preamble, a Start-bit, Data and a Stop bit. Effectively, the data is represented by four combinations of short or long pulses and short or long gaps, so that's what my software recreates - with no idea of its meaning. (If anyone can spot the encoding, I _would_ be interested to know!)

Dude 

The coding scheme used in tmps initial process is "manchester code" also known as phase encoding ,
For demodulation signal you must cut negative part of signal .
The positive part must be invert to show original data logic level ( in ook and ask level 0 means max amplitude and level 1 mean zero or low amplitude in rf signal )

* Preamble or initial pattern are case to case different depends on manufactorer

**Comment by kenwawa01** [comment]
@hornbyp 

It's not like that I didn't acknowledge. I've already ordered two things which I hope might become antennas, one is rfid antenna, second is ferrite with coil but I have to wait until I get them and then I will see what I will be able to get as result. That's why I didn't focused on this area not having "toys" for experiments.


Concerning this image, that's what I'm expecting to get, but currently I'm getting sth which is almost flat and I have problems to distinct noise from real data.

And yes I want to understand the protocol, at least up to possible extent, and not only to repeat the signal.

**Comment by ken-wawa-wawa** [comment]
I'm returning to this thread after long long time (due to very serious personal turbulences)...

Getting to the point. I recorded few examples of 125kHz programming/activation sequences with my oscilloscope (with connected coil) but not sure about the result. Unfortunately the model I have is far from being perfect when it comes about easy to use gui for saving data.

Ex. Let's try one file: https://drive.google.com/file/d/18ikjbx4ziiJFkj2p3EJThzU0kc3VNXTa/view?usp=sharing

When I analyze it with https://triq.org/pdv/ (sorry I don't know how to share the file directly from this site), surprisingly what it makes sense to me is the following configuration:

Pulses	26× 499.4 ±23.5 µs	22× 994.9 ±49.0 µs
Gaps	47× 497.6 ±24.5 µs
Periods	25× 998.0 ±44.0 µs	22× 1490.5 ±65.5 µs
Timings	74× 498.0 ±24.5 µs	22× 994.9 ±49.0 µs
DC bias (Pulse/Gap skew): 49.1%
Guessing modulation: Pulse Width Modulation with fixed gap
modulation: PWM short: 499.4 long: 994.9 sync: - gap: - reset: 628.8
RfRaw (rx): AAB10201F203E380908090809080908090809080908090809080908090809080908090809090808090809090909090808080808080808055
RfRaw (tx): AAB036020101F203E380908090809080908090809080908090809080908090809080908090809090808090809090909090808080808080808055
Bits: {48} AA AA AA A9 A0 FF

It looks like a very short activation sequence and I have very serious doubts if I'm doing it well.

To my knowledge it's exactly 125kHz signal with expected bit rate 3.9 kbit/s. I would rather expect Manchester encoding with OOK and longer activation sequence with crc at the end.

P.S. Just in case the link to the file with positive values only (if <0 then 0) https://drive.google.com/file/d/178nrmhykc3yFT2BZZ213Gg6sDw_e-lUH/view?usp=sharing However it leads to the same results in my humble opinion.

**Reply by zuckschwerdt** [reply]
Interesting! Did you try to filter the raw scope values to a OOK file, i.e. to just on-time (value != 0) and off-time (value ~= 0) pairs?

**Reply by zuckschwerdt** [reply]
I just had a peek at your shared file, it's ~8000 data points at 25 MHz, thus only 320µs long? Converted to a .cu8 to see the 125 kHz sine we just get a carrier:  [WAVE1_0.125M_25000k.cu.zip](https://github.com/user-attachments/files/20852431/WAVE1_0.125M_25000k.cu.zip) -- do you have a longer recording, a few seconds maybe?

**Reply by ken-wawa-wawa** [reply]
Thanks for help. The issue is that my osciloscoppe allows to record 8000 samples only. Moreover I don't have any clear influence on time interval between samples. It somehow calculates it basing on horizontal scale. So If I increase time interval, the signal will not be enough detailed. Hover please have a look on the second example https://drive.google.com/file/d/1asqooE13yUEXkKx1VZ08K1SZLTYy5s-p/view?usp=sharing it's 6.4ms. However it leads to the same sequence Bits: {48} AA AA AA A9 A0 FF 

What time interval between samples would be the best (if you can take only 8k samples)?

I'm thinking about buying esp32 with +  EM4095 chip and perhaps to build sth on that. Do you perhaps have any good example?

P.S. How do you convert my files to .cu ones?

**Reply by zuckschwerdt** [reply]
You will need more than 250 kHz sampling rate (https://en.wikipedia.org/wiki/Nyquist%E2%80%93Shannon_sampling_theorem), usually with a good amount of oversampling, e.g. 8x would be 2 MHz. Try maybe 500 kHz, it should give you 4 samples per sine, which should be enough to determine OOK.

Not sure about the  EM4095. You need a passive 125 kHz antenna only, not an active 125 kHz carrier. But I don't know much about that and could be wrong.

```python
#!/usr/bin/env python3

import sys, struct

filename = sys.argv[1]
filenameOut = filename + '_0.125M_25000k.cu8'

with open(filename) as file, open(filenameOut, 'wb') as out:
    for line in file:
        if line.startswith('index,'):
            break
    for line in file:
        fields = line.split(',')
        val = float(fields[1]) / 3 # assume +/-320 max
        out.write(struct.pack('2B', int(val) + 128, 0)) # I (=val) + Q (=0)
```

**Reply by ken-wawa-wawa** [reply]
Hmmm... I played more with that osciloscoppe saving and I have impression that sth is wrong. It's universal tpms activation tool and I've selected few different car models and... for all I get the same activation sequence. It doesn't seem to be possible...

I have to do it in a different way. Not only sure how...
